hphp/ppc64-asm/asm-ppc64.cpp

```diff
@@ -19,33 +19,35 @@
 #include "hphp/ppc64-asm/decoder-ppc64.h"
 #include "hphp/runtime/base/runtime-option.h"
 #include "hphp/util/trace.h"
+#include <folly/MicroSpinLock.h>
 
 TRACE_SET_MOD(asmppc64);
 
 namespace ppc64_asm {
-VMTOC::~VMTOC() {
-  FTRACE(1, "Number of values stored in TOC: {}\n",
-    std::to_string(m_last_elem_pos));
-}
 
-int64_t VMTOC::pushElem(int64_t elem) {
-  auto& map_elem = m_map[elem];
-  if (map_elem) return map_elem;
+// Lock to protect TOC when writing.
+static folly::MicroSpinLock s_TOC;
 
-  auto offset = allocTOC(static_cast<int32_t>(elem & 0xffffffff), true);
-  map_elem = offset;
-  allocTOC(static_cast<int32_t>((elem & 0xffffffff00000000) >> 32));
-  m_last_elem_pos += 2;
-  return offset;
+//////////////////////////////////////////////////////////////////////
+
+VMTOC::~VMTOC() {
+  FTRACE(1, "Number of values if 64bits stored in TOC: {}\n",
+    std::to_string(m_last_elem_pos));
 }
 
-int64_t VMTOC::pushElem(int32_t elem) {
-  auto& map_elem = m_map[elem];
-  if (map_elem) return map_elem;
+int64_t VMTOC::pushElem(int64_t elem, bool elemMayChange) {
+  int64_t offset;
+  if(elemMayChange) {
```

Comment:
Also—can you explain to me what is the consequence of not caching the offset in `m_map`?